src/iterators/optional_iterator.c

```diff
@@ -0,0 +1,112 @@
+#include "optional_iterator.h"
+
+static void OI_Free(QueryIterator *base) {
+  OptionalIterator *oi = (OptionalIterator *)base;
+  if (oi->child) {
+    oi->child->Free(oi->child);
+  }
+  // Only free our virtual result, not the child's result
+  if (oi->virt) {
+    IndexResult_Free(oi->virt);
+  }
+  rm_free(base);
+}
+
+static size_t OI_NumEstimated(QueryIterator *base) {
+  OptionalIterator *oi = (OptionalIterator *)base;
+  return oi->maxDocId;
+}
+
+static void OI_Rewind(QueryIterator *base) {
+  OptionalIterator *oi = (OptionalIterator *)base;
+  base->atEOF = false;
+  base->lastDocId = 0;
+  oi->virt->docId = 0;
+  if (oi->child) {
+    oi->child->Rewind(oi->child);
+  }
+}
+
+
+
+// SkipTo for OPTIONAL iterator - Non-optimized version.
+// Skip to a specific docId. If the child has a hit on this docId, return it.
+// Otherwise, return a virtual hit.
+static IteratorStatus OI_SkipTo_NotOptimized(QueryIterator *base, t_docId docId) {
+  RS_ASSERT(docId > 0);
+  RS_ASSERT(docId > base->lastDocId);
+  OptionalIterator *oi = (OptionalIterator *)base;
+
+  if (docId > oi->maxDocId || base->atEOF) {
+    base->atEOF = true;
+    return ITERATOR_EOF;
+  }
+
+  if (docId > oi->child->lastDocId) {
+    IteratorStatus rc = oi->child->SkipTo(oi->child, docId);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+  }
+
+  if (docId == oi->child->lastDocId) {
+    // Has a real hit on the child iterator
+    base->current = oi->child->current;
+    base->current->weight = oi->weight;
+  } else {
+    // Virtual hit
+    oi->virt->docId = docId;
+    base->current = oi->virt;
+  }
+  // Set the current ID
+  base->lastDocId = docId;
+  return ITERATOR_OK;
+}
+
+// Read from an OPTIONAL iterator - Non-Optimized version.
+static IteratorStatus OI_ReadSorted_NotOptimized(QueryIterator *base) {
+  OptionalIterator *oi = (OptionalIterator *)base;
+  if (base->atEOF || base->lastDocId >= oi->maxDocId) {
+    base->atEOF = true;
+    return ITERATOR_EOF;
+  }
+
+  // Point to next doc
+  base->lastDocId++;
+
+  if (base->lastDocId > oi->child->lastDocId) {
+    IteratorStatus rc = oi->child->Read(oi->child);
+    if (rc == ITERATOR_TIMEOUT) return rc;
+  }
+
+  if (base->lastDocId == oi->child->lastDocId) {
+    base->current = oi->child->current;
+    base->current->weight = oi->weight;
+  } else {
+    oi->virt->docId = base->lastDocId;
+    base->current = oi->virt;
+  }
+  return ITERATOR_OK;
+}
+
+// Create a new OPTIONAL iterator - Non-Optimized version.
+QueryIterator *IT_V2(NewOptionalIterator)(QueryIterator *it, t_docId maxDocId, size_t numDocs, double weight) {
+  assert(it != NULL);
```

Comment:
On a second look, this may happen (for example, when searching for `~foo` but no document has `foo` so there is no inverted index for it)
I suggest we return a wildcard iterator in this case.
We have to check the implications on `FT.PROFILE` and `FT.EXPLAIN` when we start using the new iterators

Reply:
alternativly, use `RS_ASSERT` and we will have to handle this case when using the API